pkg/resources/envoy/envoy.go

```diff
@@ -53,41 +67,87 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
-				externalListenerResources = append(externalListenerResources,
-					r.service,
-					r.configMap,
-					r.deployment,
-				)
 
-				if r.KafkaCluster.Spec.EnvoyConfig.GetDistruptionBudget().DisruptionBudget.Create {
-					externalListenerResources = append(externalListenerResources, r.podDisruptionBudget)
+				for _, res := range externalListenerResources {
+					o := res(log, eListener, ingressConfig, name, defaultControllerName)
+					err := k8sutil.Reconcile(log, r.Client, o, r.KafkaCluster)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		} else {
+			// Cleaning up unused envoy resources when ingress controller is not envoy or externalListener access method is not LoadBalancer
+			deletionCounter := 0
+			ctx := context.Background()
+			envoyResourcesGVK := []schema.GroupVersionKind{
+				{
+					Version: corev1.SchemeGroupVersion.Version,
+					Group:   corev1.SchemeGroupVersion.Group,
+					Kind:    reflect.TypeOf(corev1.Service{}).Name(),
+				},
+				{
+					Version: corev1.SchemeGroupVersion.Version,
+					Group:   corev1.SchemeGroupVersion.Group,
+					Kind:    reflect.TypeOf(corev1.ConfigMap{}).Name(),
+				},
+				{
+					Version: appsv1.SchemeGroupVersion.Version,
+					Group:   appsv1.SchemeGroupVersion.Group,
+					Kind:    reflect.TypeOf(appsv1.Deployment{}).Name(),
+				},
+				{
+					Version: policyv1.SchemeGroupVersion.Version,
+					Group:   policyv1.SchemeGroupVersion.Group,
+					Kind:    reflect.TypeOf(policyv1.PodDisruptionBudget{}).Name(),
+				},
+			}
+			var envoyResources unstructured.UnstructuredList
+			for _, gvk := range envoyResourcesGVK {
+				envoyResources.SetGroupVersionKind(gvk)
+
+				if err := r.List(ctx, &envoyResources, client.InNamespace(r.KafkaCluster.GetNamespace()),
+					client.MatchingLabels(labelsForEnvoyIngressWithoutEListenerName(r.KafkaCluster.Name))); err != nil {
+					return errors.Wrap(err, "error when getting list of envoy ingress resources for deletion")
 				}
-				for name, ingressConfig := range ingressConfigs {
-					if !util.IsIngressConfigInUse(name, defaultControllerName, r.KafkaCluster, log) {
+
+				for _, removeObject := range envoyResources.Items {
+					if !strings.Contains(removeObject.GetLabels()[util.ExternalListenerLabelNameKey], eListener.Name) ||
+						util.ObjectManagedByClusterRegistry(&removeObject) ||
+						!removeObject.GetDeletionTimestamp().IsZero() {
 						continue
 					}
-
-					for _, res := range externalListenerResources {
-						o := res(log, eListener, ingressConfig, name, defaultControllerName)
-						err := k8sutil.Reconcile(log, r.Client, o, r.KafkaCluster)
-						if err != nil {
-							return err
-						}
+					if err := r.Delete(ctx, &removeObject); client.IgnoreNotFound(err) != nil {
+						return errors.Wrap(err, "error when removing envoy ingress resources")
 					}
+					deletionCounter++
 				}
 			}
+			if deletionCounter > 0 {
+				log.Info(fmt.Sprintf("Removed '%d' resources for envoy ingress", deletionCounter))
+			}
```

Comment:
**Question**: Wouldn't it be better to put a log message in the place of `deleteCounter++` stating `deleted envoy ingress resource %s` where `%s` is the name of the object?
This way we will see `removed 5 resources for envoy ingress`, not sure if that is more useful than seeing `removed resources for envoy ingress` vs. the more informative listing of removed resources.
I can be convinced we don't need the resource level info, just curious.

Same for the Istio version.

Reply:
Then it would be 5 remove message in case envoy.  I think it is a good idea to do this with lower log level like: log.V(1).Info(...) What do you think?